server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -573,12 +578,43 @@ public void addAllocFailuresResetListenerTo(ClusterService clusterService) {
         });
 
         clusterService.addListener((changeEvent) -> {
-            if (changeEvent.nodesAdded() && changeEvent.state().getRoutingNodes().hasAllocationFailures()) {
+            if (shouldResetAllocationFailures(changeEvent)) {
                 taskQueue.submitTask("reset-allocation-failures", (e) -> { assert MasterService.isPublishFailureException(e); }, null);
             }
         });
     }
 
+    /**
+     *  We should reset allocation/relocation failure count to allow further retries when:
+     *
+     *  1. A new node joins the cluster.
+     *  2. A node shutdown metadata is added that could lead to a node being removed or replaced in the cluster.
+     *
+     * Note that removing a non-RESTART shutdown metadata from a node that is still in the cluster is treated similarly and
+     * will cause resetting the allocation/relocation failures.
+     */
+    private boolean shouldResetAllocationFailures(ClusterChangedEvent changeEvent) {
+        final var clusterState = changeEvent.state();
+        boolean hasAllocationFailures = clusterState.getRoutingNodes().hasAllocationFailures();
+        boolean hasRelocationFailures = clusterState.getRoutingNodes().hasRelocationFailures();
+        var shutdownEventAffectsAllocation = false;
+        final var nodes = clusterState.nodes();
+        final var nodeShutdowns = clusterState.metadata().nodeShutdowns();
+        // If we remove a shutdown marker from a node, but it is still in the cluster, we could re-attempt failed relocations/allocations.
+        shutdownEventAffectsAllocation = processedNodeShutdowns.stream()
+            .anyMatch(nodeId -> nodeShutdowns.contains(nodeId) == false && nodes.get(nodeId) != null);
```

Comment:
This will cover restarted nodes as well. I think it's ok since it is covered by `nodesAdded` anyway. It may lead to reset a few more times:
1. On shutdown metadata (either restart or sigterm for serverless azure cluster)
2. On node join after restart
3. On shutdown metadata removal (if in a later cluster state update)

Still might be ok. Just to be explicit.

Reply:
We explicitly avoid reset when the RESTART metadata is applied. When the node rejoins, yes, there will be a reset. We could explicitly prevent it, but not sure it is worth it. Removal of RESTART shouldn't cause a reset if the node is in the cluster. I think it is tested as well.

Reply:
I should have been clear, by "restart", I meant the Azure specific behaviour where the node still goes through the sigterm shutdown record lifecyle while restarted by the Azure "auto-repair" feature. When it happens, we will reset the counter a few more times which should be ok.